File: src/lib/paymentService.ts
import { getFunctions, httpsCallable } from 'firebase/functions';
import { TEST_PAYMENT_DATA } from '../config/testPaymentData';

export interface PaymentRequest {
  method: 'card' | 'instapay';
  amount: number;
  userId?: string;
  orderId?: string;
  cardDetails?: {
    number: string;
    expiryDate: string;
    cvv: string;
  };
  instapayNumber?: string;
}

export interface PaymentResponse {
  success: boolean;
  transactionId?: string;
  paymentId?: string;
  status?: string;
  cardType?: string;
  method?: string;
  amount?: number;
  message?: string;
  error?: {
    code: string;
    message: string;
  };
}

export interface PaymentLog {
  operation: string;
  timestamp: Date;
  userRole: string;
  userId: string;
  paymentMethod: string;
  amount: number;
  status: 'success' | 'error' | 'pending';
  transactionId?: string;
  errorMessage?: string;
  debugTrace?: string;
}

class PaymentService {
  private functions = getFunctions();

  /**
   * Process payment using unified API
   */
  async processPayment(paymentRequest: PaymentRequest, userRole: string, userId: string): Promise<PaymentResponse> {
    const log: PaymentLog = {
      operation: 'payment_processing',
      timestamp: new Date(),
      userRole,
      userId,
      paymentMethod: paymentRequest.method,
      amount: paymentRequest.amount,
      status: 'pending',
      debugTrace: `Payment initiated for ${paymentRequest.method} - Amount: $${paymentRequest.amount}`
    };

    try {
      let response: PaymentResponse;

      if (paymentRequest.method === 'card') {
        response = await this.processCardPayment(paymentRequest);
      } else if (paymentRequest.method === 'instapay') {
        response = await this.processInstapayPayment(paymentRequest);
      } else {
        throw new Error('Invalid payment method');
      }

      // Update log with success
      log.status = 'success';
      log.transactionId = response.transactionId;
      log.debugTrace = `Payment successful - Transaction ID: ${response.transactionId}`;

      // Log the payment
      await this.logPayment(log);

      return response;

    } catch (error: any) {
      // Update log with error
      log.status = 'error';
      log.errorMessage = error.message;
      log.debugTrace = `Payment failed - Error: ${error.message}`;

      // Log the error
      await this.logPayment(log);

      return {
        success: false,
        error: {
          code: error.code || 'payment_failed',
          message: error.message || 'Payment processing failed'
        }
      };
    }
  }

  /**
   * Process card payment
   */
  private async processCardPayment(paymentRequest: PaymentRequest): Promise<PaymentResponse> {
    if (!paymentRequest.cardDetails) {
      throw new Error('Card details are required for card payment');
    }

    const processCardPayment = httpsCallable(this.functions, 'processCardPayment');

    const result = await processCardPayment({
      cardNumber: paymentRequest.cardDetails.number,
      expiryDate: paymentRequest.cardDetails.expiryDate,
      cvv: paymentRequest.cardDetails.cvv,
      amount: paymentRequest.amount,
      userId: paymentRequest.userId,
      orderId: paymentRequest.orderId
    });

    return result.data as PaymentResponse;
  }

  /**
   * Process Instapay payment
   */
  private async processInstapayPayment(paymentRequest: PaymentRequest): Promise<PaymentResponse> {
    if (!paymentRequest.instapayNumber) {
      throw new Error('Instapay number is required for Instapay payment');
    }

    const processInstapayPayment = httpsCallable(this.functions, 'processInstapayPayment');

    const result = await processInstapayPayment({
      instapayNumber: paymentRequest.instapayNumber,
      amount: paymentRequest.amount,
      userId: paymentRequest.userId,
      orderId: paymentRequest.orderId
    });

    return result.data as PaymentResponse;
  }

  /**
   * Log payment operation
   */
  private async logPayment(log: PaymentLog): Promise<void> {
    try {
      // This will be handled by Cloud Functions automatically
      // The log will be written to Firestore logs collection
      console.log('Payment Log:', log);
    } catch (error) {
      console.error('Failed to log payment:', error);
    }
  }

  /**
   * Validate test payment credentials
   */
  validateTestCredentials(paymentRequest: PaymentRequest): boolean {
    if (paymentRequest.method === 'card') {
      if (!paymentRequest.cardDetails) return false;

      const { number, expiryDate, cvv } = paymentRequest.cardDetails;
      const cleanNumber = number.replace(/\s+/g, '').replace(/-/g, '');
      const cleanExpiry = expiryDate.replace(/\//g, '');

      // Check against test cards
      const isVisa = cleanNumber === TEST_PAYMENT_DATA.cards.visa.number.replace(/\s+/g, '') &&
                    cleanExpiry === TEST_PAYMENT_DATA.cards.visa.expiry.replace(/\//g, '') &&
                    cvv === TEST_PAYMENT_DATA.cards.visa.cvv;

      const isMasterCard = cleanNumber === TEST_PAYMENT_DATA.cards.mastercard.number.replace(/\s+/g, '') &&
                          cleanExpiry === TEST_PAYMENT_DATA.cards.mastercard.expiry.replace(/\//g, '') &&
                          cvv === TEST_PAYMENT_DATA.cards.mastercard.cvv;

      return isVisa || isMasterCard;

    } else if (paymentRequest.method === 'instapay') {
      return paymentRequest.instapayNumber === TEST_PAYMENT_DATA.instapay.number;
    }

    return false;
  }

  /**
   * Get test payment credentials
   */
  getTestCredentials() {
    return {
      cards: TEST_PAYMENT_DATA.cards,
      instapay: TEST_PAYMENT_DATA.instapay
    };
  }
}

export const paymentService = new PaymentService();
export default paymentService; 

File: lib/services/payment_service.dart
import 'package:cloud_functions/cloud_functions.dart';

class PaymentService {
  final FirebaseFunctions _functions = FirebaseFunctions.instance;

  // Process card payment (Visa/MasterCard)
  Future<Map<String, dynamic>> processCardPayment({
    required String cardNumber,
    required String expiryDate,
    required String cvv,
    required double amount,
    String? userId,
    String? orderId,
  }) async {
    try {
      final callable = _functions.httpsCallable('processCardPayment');
      final result = await callable.call({
        'cardNumber': cardNumber,
        'expiryDate': expiryDate,
        'cvv': cvv,
        'amount': amount,
        'userId': userId,
        'orderId': orderId,
      });
      
      return result.data as Map<String, dynamic>;
    } catch (e) {
      throw 'Error processing card payment: ${e.toString()}';
    }
  }

  // Process Instapay payment
  Future<Map<String, dynamic>> processInstapayPayment({
    required String instapayNumber,
    required double amount,
    String? userId,
    String? orderId,
  }) async {
    try {
      final callable = _functions.httpsCallable('processInstapayPayment');
      final result = await callable.call({
        'instapayNumber': instapayNumber,
        'amount': amount,
        'userId': userId,
        'orderId': orderId,
      });
      
      return result.data as Map<String, dynamic>;
    } catch (e) {
      throw 'Error processing Instapay payment: ${e.toString()}';
    }
  }










  // Validate card number format
  static bool isValidCardNumber(String cardNumber) {
    // Remove spaces and dashes
    final cleanNumber = cardNumber.replaceAll(RegExp(r'[\s-]'), '');
    
    // Check if it's numeric and has correct length
    if (!RegExp(r'^\d+$').hasMatch(cleanNumber)) return false;
    if (cleanNumber.length < 13 || cleanNumber.length > 19) return false;
    
    return true;
  }

  // Validate expiry date format
  static bool isValidExpiryDate(String expiryDate) {
    final regex = RegExp(r'^(0[1-9]|1[0-2])\/([0-9]{2})$');
    if (!regex.hasMatch(expiryDate)) return false;
    
    final parts = expiryDate.split('/');
    final month = int.parse(parts[0]);
    final year = int.parse('20${parts[1]}');
    
    final now = DateTime.now();
    final expiry = DateTime(year, month);
    
    return expiry.isAfter(now);
  }

  // Validate CVV
  static bool isValidCVV(String cvv) {
    return RegExp(r'^\d{3,4}$').hasMatch(cvv);
  }

  // Validate Instapay number
  static bool isValidInstapayNumber(String instapayNumber) {
    return RegExp(r'^01[0-9]{9}$').hasMatch(instapayNumber);
  }
}
